src/Adapters/Auth/index.js

```diff
@@ -59,21 +61,60 @@ const providers = {
   microsoft,
   keycloak,
   ldap,
+  webauthn,
 };
 
-function authDataValidator(adapter, appIds, options) {
-  return function (authData) {
-    return adapter.validateAuthData(authData, options).then(() => {
-      if (appIds) {
-        return adapter.validateAppId(appIds, authData, options);
+function authDataValidator(provider, adapter, appIds, options) {
+  return async function (authData, req, user) {
+    if (appIds && typeof adapter.validateAppId === 'function') {
+      await Promise.resolve(adapter.validateAppId(appIds, authData, options, req, user));
+    }
+    if (typeof adapter.validateAuthData === 'function') {
+      return adapter.validateAuthData(authData, options, req, user);
+    } else if (
+      typeof adapter.validateSetUp === 'function' &&
+      typeof adapter.validateLogin === 'function' &&
+      typeof adapter.validateUpdate === 'function'
+    ) {
+      // When masterKey is detected, we should trigger a logged in user
+      const isLoggedIn =
+        (req.auth.user && user && req.auth.user.id === user.id) || (user && req.auth.isMaster);
+      let hasAuthDataConfigured = false;
+
+      if (user && user.get('authData') && user.get('authData')[provider]) {
+        hasAuthDataConfigured = true;
+      }
+
+      if (isLoggedIn) {
+        // User is updating their authData
+        if (hasAuthDataConfigured) {
+          return adapter.validateUpdate(authData, options, req, user);
+        }
+        // Let's setup if the user does not have the provider configured
+        return adapter.validateSetUp(authData, options, req, user);
+      }
+
+      // Not logged in and authData is configured on the user
+      if (hasAuthDataConfigured) {
+        return adapter.validateLogin(authData, options, req, user);
       }
-      return Promise.resolve();
-    });
+
+      // User not logged in and the provider is not set up, for example when a new user
+      // signs up or an existing user uses a new auth provider
+      return adapter.validateSetUp(authData, options, req, user);
+    }
+    throw new Parse.Error(
+      Parse.Error.OTHER_CAUSE,
+      'Adapter not ready, need to implement validateAuthData or (validateSetUp, validateLogin, validateUpdate)'
```

Comment:
`Adapter is not configured. Implement validateAuthData, or validateSetUp, validateLogin and validateUpdate`

Reply:
> Implement validateAuthData, or validateSetUp, validateLogin and validateUpdate

Sounds confusing, and / or what?

Reply:
I like `Adapter is not configured. Implement either validateAuthData or all of the following: validateSetUp, validateLogin and validateUpdate`

Reply:
fixed